Introduce kind

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Related to #20158 

#### Brief description of what is fixed or changed

1. `core/kind` module is introduced.
2. `Basic` has `UndefinedKind` as default kind.
3. Every classes in `core/numbers` module, `int`, `float` have `NumberKind.
4.  NDimArrays have parametrisable `ArrayKind`. Matrices have parametrisable `MatrixKind`.
5. Boolean expressions have `BooleanKind`.
6. Following classes support kind inference:
    - `Add`
    - `Symbol`
    - `Integral`
    - `Tr`
7. Documentation is made in `docs/src/special_topics`.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - Kind classification of objects is introduced. This feature is experimental, and can be replaced or deleted in the future.
<!-- END RELEASE NOTES -->

JSS95 commented:
@oscarbenjamin 
I made some rudimentary implementation of your idea. Currently, `NumberType` and `MatrixType` are introduced, and `Add` and `Mul` supports typeclass inference.
```python
>>> from sympy.core.core import typeclass
>>> from sympy import MatrixSymbol

>>> typeclass(2)
NumberSymbol
>>> typeclass(2*MatrixSymbol('A',2,3))
MatrixSymbol
```

This will need many trials and reviews. If you have any suggestion, please let me know and I am ready to follow.

oscarbenjamin commented:
What is the advantage of using single dispatch here rather than a method/attribute?

I was expecting something like:
```python
def typeclass(obj):
    return obj._typeclass

class Number:
    _typeclass = NumberType

class MatrixExpr:
     _typeclass = MatrixType

class Set:
    _typeclass = SetType

class Add:
    def __new__(cls, args):
        if len(set(map(typeclass, args))) != 1:
            raise TypeError
        ...
        obj._typeclass = typeclass(args[0])
        return obj
```
This way the `typeclass` function is mostly just retrieving a class/instance attribute which is fast. We could also have a property for accessing this defined on Basic:
```python
class Basic:
    _typeclass = UnknownType

    @property
    def typeclass(self):
        return self._typeclass
```

Usually the advantage of single dispatch is that you want to be able to overload a function for a particular class without modifying the class itself. That's useful in some situations but in the case of the typeclass it always needs to be defined so we would have to add the dispatch immediately after the class definition anyway.

JSS95 commented:
@oscarbenjamin 
I changed the design as your suggestion. BTW how shall we do with `Add`, `Mul`, `Pow` etc? I still think that we can get rid of `MatAdd`, `MatMul` with careful redesigning of the classes...

oscarbenjamin commented:
> BTW how shall we do with `Add`, `Mul`, `Pow` etc? I still think that we can get rid of `MatAdd`, `MatMul` and unify everything to `Add` and `Mul` with careful redesigning of the classes...

I think that this is the tricky part. For now I would say let's not try to get rid of MatAdd etc as that would add a lot of churn and it's not clear what the benefit would be now that we have those classes. If we can use this to help eliminate the post-processors then that's good.

It's important not to get too carried away with something like this. For now we should keep the changes as small as possible but make sure that the basic structure is something that can work. In particular we need to ensure:

1. `typeclass` always returns quickly and it reliably handles the difference between e.g. number and matrix (including in nested expressions) for meaningful expressions.
2. We have a clear way to get e.g. the shape of any MatrixType expression.
3. There is clear documentation explaining what the definitions of the typeclasses are in a way that can be referenced from other parts of the docs.
4. Any changes are backwards compatible for downstream code that hasn't defined any typeclasses or that is mixing them in ways we might consider to be inconsistent but are currently allowed.

Potentially for `Add` it makes sense to require all terms to have the same typeclass but that wouldn't make sense for `Mul` and `Pow`. When I say "require" we can't make this an error because that would not be backwards compatible. Rather we need to document what should be allowed in a way that is clear to users, contributors, etc. Existing code should continue to work and it is premature even to give a deprecation warning at this stage. We would need the typeclasses to exist in a few sympy releases at least before deprecating anything.

We could say for e.g. `Mul` that there can be a mixture of `NumberType` and one other typeclass say `XType` which means that the typeclass of the `Mul` itself is `XType`. Potentially the logic here is the same as for the `add`, `mul` dispatched functions.

For `Pow` the typeclass should be the same as the base if the exponent is `NumberType`. I don't know if there are sensible examples of `Pow` with a non-`NumberType` exponent currently in sympy (or dependent libraries). In principle you can have e.g. `M**N` where `M` and `N` are both matrices and the result is a matrix. I don't know of any meaningful examples where the typeclass of a `Pow` could differ from the typeclass of the base though.

JSS95 commented:
Agreed. I am aiming to restrict the scope of this PR to rudimentary implementation.

JSS95 commented:
> If we can use this to help eliminate the post-processors then that's good.

Exactly. If we no longer need `MatAdd`, `VectorAdd`, etc, then we won't need postprocessors and dispatched `add`, `mul` functions as well.

> Potentially for Add it makes sense to require all terms to have the same typeclass

Sounds true, since addition is group operator. I changed the design that way in the latest commit.

> Potentially the logic here is the same as for the add, mul dispatched functions.

Yes. I implemented `TypeclassDispatcher` in `core/operator` module for this purpose.

> I don't know of any meaningful examples where the typeclass of a Pow could differ from the typeclass of the base though.

Mathematically, we have [matrix exponential](https://en.wikipedia.org/wiki/Matrix_exponential) so probably we should make `Pow(2, M)` (where `M` is matrix) have `MatrixType`.

oscarbenjamin commented on sympy/core/typeclass.py:
```diff
@@ -0,0 +1,59 @@
+        return "MatrixType"
+
+
+MatrixType = _MatrixType()
```
Maybe this should be defined in the matrices module.

oscarbenjamin commented:
> Mathematically, we have [matrix exponential](https://en.wikipedia.org/wiki/Matrix_exponential) so probably we should make `Pow(2, M)` (where `M` is matrix) have `MatrixType`.

Oh, that's true. We do already have the matrix exponential as well:
```python
In [12]: exp(Matrix([[0, -1], [1, 0]]))                                                                                           
Out[12]: 
⎡cos(1)  -sin(1)⎤
⎢               ⎥
⎣sin(1)  cos(1) ⎦
```

JSS95 commented:
@oscarbenjamin 
I equipped every classes in `core` with typeclass. I think we can now focus on enhancing the design instead of adding something more in this PR.

JSS95 commented:
Plus, I found an inconsistent behavior related to matrix and array derivation.

```python
>>> v = Matrix([x,y])
>>> type(v.diff(x))
sympy.matrices.dense.MutableDenseMatrix
>>> type(x.diff(v))
sympy.matrices.immutable.ImmutableDenseMatrix
>>> type(v.diff(v))
sympy.tensor.array.dense_ndim_array.MutableDenseNDimArray
```

Since the derivation of matrix by matrix returns `NDimArray`, I had to give `ArrayType` to unevaluated derivatives involving any matrix.

In my opinion, this is the class design failure rather than typeclass's problem. Matrix is 2 dimensional array, which is a special case of N dimensional array. They are important enough to deserve their own `matrices` module, but conceptually we should make them subclass of `NDimArray` and guarantee full compatibility between these two.

Fortunately, we do not need to worry about this now because unevaluated matrix derivative hasn't been supported anyway.

eric-wieser commented on doc/src/gotchas.rst:
```diff
@@ -758,6 +758,103 @@ that you want, and they will all be evaluated according to the function.
+typeclass
+---------
+
+Typeclass [1]_ indicates what mathematical object (in broad sense) does the
+expression represent. You can retrieve the typeclass of expression with
+`sympy.core.typeclass()` function.
+
+    >>> from sympy.core import typeclass
```
Having spent the last few months working with "typeclasses" in lean, where they resemble haskell - I don't think this is the correct meaning of the word, and wikipedia confirms this with the phrasing "belongs to a typeclass".

For example, a number type would likely belong to all of `Ordered`, `HasAdd`, and  `Hashable` typeclasses.
Your implementation has each type assign itself a _single_ `typeclass`. This is not how typeclasses work at all - instead, they consist of external bits of code registering a variety of typeclasses (aking to multiple dispatch) for a variety of types.

[PEP544](https://www.python.org/dev/peps/pep-0544/) looks to me like a reasonable approximation of typeclasses for static-typing of typeclasses, while duck-typing is already the equivalent for implementing typeclasses.

I think you need to come up with a different name.

JSS95 replied:
@oscarbenjamin Any idea?

oscarbenjamin replied:
I guess any name we go with would clash with something.

All the names have already been taken though: category, class, type, etc. Presumably that's why other people already came up with "typeclass".

Okay so synonyms from Macmillan:
```
Categories and types
SYNONYMS
type NOUN
a group of people or things with similar qualities or features that make them different from other groups

category NOUN
a group of people or things that have similar qualities

kind NOUN
a type of person or thing

sort NOUN
a group of people or things with the same qualities or features

variety NOUN
a type of thing

classification NOUN
a group into which someone or something is put because of the features that they have

grouping NOUN
a set of people or things that are considered as a group

taxonomy NOUN
a system for organizing similar things into groups

class NOUN
a group of things, animals, or people with similar features or qualities

cross section NOUN
a group that contains an example of all or most types of people or things
```
https://www.macmillandictionary.com/thesaurus-category/british/categories-and-types

From the definitions given it would naturally seem that "type" is most suited but it already has a meaning in Python where both `type` and `Type` are already used. It also most closely matches e.g. type theory:
https://en.wikipedia.org/wiki/Type_theory
Maybe referring to something like type theory is a mistake though. The differences might be more distracting than the similarities (as for typeclasses).

Synonyms of "type"  are "kind" and "variety". Since "variety" is already used in maths we could consider "kind". Then we have "NumberKind", "MatrixKind" and
```python
def sin(expr):
    """sine function

    parameters
    ==========

    expr: NumberKind
        Any complex number
    """
    if kind(expr) is not NumberKind:
        raise TypeError
```
It seems

JSS95 replied:
I changed the naming to `kind`.

asmeurer commented:
I agree that "kind" is the best name here. "Type" would actually be the best, but there's too much potential for confusion with the Python meaning of "type" (same with "class", which has the further disadvantage that it can't be used as a variable name).

oscarbenjamin commented:
> I agree that "kind" is the best name here.

Excellent. It is well known that naming things is hard. Hopefully we've got this name nailed!

oscarbenjamin commented:
> Since the derivation of matrix by matrix returns `NDimArray`, I had to give `ArrayType` to unevaluated derivatives involving any matrix.
> 
> In my opinion, this is the class design failure rather than typeclass's problem.

I tend to agree. I have found that I don't quite understand how to formalise the way that matrix/array differentiation are mixed in a way that seems rigorous. I was recently discussing this with @Upabjojr in https://github.com/sympy/sympy/pull/20072#issuecomment-691951014

JSS95 commented:
@oscarbenjamin 
From Upapjojr's comment you linked:

> * classes inheriting MatrixCommon: explicit-valued matrices (all elements of the matrix are expressed),
> * classes inheriting MatrixExpr but not MatrixCommon: implicit-valued matrices (i.e. the matrix is just a symbol, no elements are shown),
> * MatrixElement class: I'd call it implicit-valued (as elements are not shown), explicit-indexed (as indices are shown, unlike MatrixExpr),
> * classes inheriting NDimArray: like explicit matrices, but with N-axes (matrices always have 2 axes),
> * IndexedBase and Indexed objects: generic indexed objects, implicit-valued explicit-index,
> classes in sympy.codegen.array_utils: implicit-valued implicit-indexed arrays, still experimental.
> * sympy.tensor.tensor: abstract index notation for tensorial expressions. This is reserved for applications in theoretical physics.

Current design seems very complicated and unintuitive. I think the class design should be like:

1. We should have classes for explicit-valued and implicit-valued n-dimensional arrays, all subclassing `NDimArray`. For example, something like `ArraySymbol` should exist.

2. Explicit-valued matrix should subclass explicit-valued n-dimensional array. Similar for implicit-valued matrix.

3. `MatrixElement` should subclass `ArrayElement`, and perhaps this should subclass `Indexed`.

oscarbenjamin commented:
> 5\. `Tuple`, `tuple` and `list` have `SequenceKind`.

I hadn't considered the idea that there could be a kind for non-Basic types. I'm not sure if that's a good idea.

For tuple I guess it makes sense because tuple sympifies to Tuple. We don't have a Basic equivalent of list though. I had imagined that `kind` was a function that applied only to Basic and would be used only after sympification. There are also some other non-Basic types like Matrix that would probably need a kind. Matrix simpifies to ImmutableMatrix though which perhaps makes sense like tuple/Tuple.

oscarbenjamin commented on doc/src/gotchas.rst:
```diff
@@ -758,6 +758,103 @@ that you want, and they will all be evaluated according to the function.
+Kind indicates what mathematical object does the expression represent.
+You can retrieve the kind of expression with `sympy.core.kind()` function.
+
+    >>> from sympy.core import kind
```
If we have a kind function then it should be at top-level import like `from sympy import kind`

oscarbenjamin commented on sympy/matrices/expressions/matpow.py:
```diff
@@ -138,3 +145,7 @@ def _eval_derivative_matrix_lines(self, x):
+
+
+Pow._kind_dict[(NumberKind, MatrixKind)] = MatrixKind
+Pow._kind_dict[(MatrixKind, NumberKind)] = MatrixKind
```
Would it make more sense to give MatPow a distinct `_kind_dict`?

This modifies the `_kind_dict` attribute of `Pow`...

JSS95 replied:
`MatPow` is not subclass of Pow.

oscarbenjamin commented:
I think that the broad idea here looks good.

I wonder if we also need an `UnknownExprKind` to distinguish from arbitrary `UnknownKind`.

There needs to be clear documentation that defines precisely what kinds of object each kind represents. A good starting point for that would be in the docstring of the `Kind` classes. That way you can do `help(kind(obj))`. The docs should define exhaustively what objects have a given kind and what operations (e.g. `shape`, `+` etc) are possible with them.

What does anyone else think about this?

@asmeurer @sylee957 @eric-wieser

sylee957 commented:
Is there any better way to centralize the tests about `Kind` than spreading everything over?
I think that a similar logic like `test_args` can be used to check whether all the sympy classes have correct logics for types, or at least  have implemented tests for it.
 I don't think that it's easy to control whether `kind` works or not correctly in this implementation, because it still gets loosely inherited by class inheritance, and let's say if someone makes a new class that inherits `Matrix` or `Add` and makes some nonsensical class and haven't defined `kind` on its own, `kind` function would still refer to the superclass

eric-wieser commented:
It's not very clear to me whether this is trying to solve the right problem.

To take an example, this makes it so that you can check if `ix = Integral(x*x, x)` is a matrix when `x` is.
The problem is, `isinstance(ix, MatrixBase)` is still false, so I can't do things like `ix.shape` or `ix.T`.
It's not clear to me what value asking for the kind of the object has if you still can't actually perform the usual operations for that kind.

My feeling is that this would be better solved by a `MatrixInterval` type, or perhaps a runtime-generated `Interval[MatrixExpr]` type, which `Interval` returns if its arguments are matrices (ideally via the type of dispatch you've explored elsewhere). I'd be much more comfortable building the behavior we need into the python class heirarchy and not our own parallel "kind" system.

oscarbenjamin commented:
> It's not clear to me what value asking for the kind of the object has if you still can't actually perform the usual operations for that kind.

Very often in sympy the operations just consist of building an expression tree and nothing actually needs to be done e.g.:
```python
In [2]: M = Matrix([[1, 2], [3, 4]])                                                                                              

In [3]: sin(M)                                                                                                                    
Out[3]: 
   ⎛⎡1  2⎤⎞
sin⎜⎢    ⎥⎟
   ⎝⎣3  4⎦⎠
```
If we know that the object mathematically represents a matrix then we know where it can be used in the expression tree. We don't necessarily need to do anything else. Knowing that it is a matrix makes it a valid argument for symbolic functions that accept matrices e.g.:
```python
In [5]: Integral(M, t)                                                                                                            
Out[5]: 
⌠          
⎮ ⎡1  2⎤   
⎮ ⎢    ⎥ dt
⎮ ⎣3  4⎦   
⌡          

In [6]: Trace(Integral(M, t))                                                                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-316cfa42acf5> in <module>
----> 1 Trace(Integral(M, t))

~/current/sympy/sympy/sympy/matrices/expressions/trace.py in __new__(cls, mat)
     31 
     32         if not mat.is_Matrix:
---> 33             raise TypeError("input to Trace, %s, is not a matrix" % str(mat))
     34 
     35         if not mat.is_square:

TypeError: input to Trace, Integral(Matrix([
[1, 2],
[3, 4]]), t), is not a matrix
```

As I explained the issue https://github.com/sympy/sympy/issues/20158#issuecomment-735230658 the most common uses of kind would be checking for invalid operations.

Of course we will also need to be able to get the shape of expressions that are of matrix kind and we can use a dispatch system to make a `shape` function that can do that. Other operations like `obj.T` would be `Transpose(obj)` here because we are working symbolically.

> My feeling is that this would be better solved by a `MatrixInterval` type

I presume you mean `MatrixIntegral`. The problem is that we have a Cartesian product of possibilities of symbolic functions that can contain different kinds of objects and also the different kinds of objects themselves e.g.:
```
{Number
Matrix
Vector
Quaternion
Tensor
...}
x
{
Mul
Add
Pow
Integral
Sum
Piecewise
Equality
...}
```
The current approach is to add classes like `MatAdd`, `MatPow` etc but we can't reasonably scale that up to everything because the number of classes needed is *quadratic*. We could have runtime generated classes but then that wouldn't actually solve the problem of how to get the shape of a matrix so it's not really any better.

I think that this PR is really codifying something that already happens in sympy: expressions of any given Python-type represent different kinds of objects depending on their args. The difference is that here we can actually check what that kind of object is. The other problem we have is about documentation and definitions. It's too late to give definitions for things like `Expr` but it's not too late to give a precise definition of `NumberKind`.

jksuom commented:
Are we expecting that all instances of certain classes are of the same "kind", like all `Number`s of `NumberKind`? If that is so, then could we make `Number` a subclass of an abstract  `NumberKind` superclass. It could be tested by `isinstance(obj, NumberKind)`.

oscarbenjamin commented:
> Are we expecting that all instances of certain classes are of the same "kind", like all `Number`s of `NumberKind`?

Yes, although there would also be classes with instances that are not all of the same kind.

> If that is so, then could we make `Number` a subclass of an abstract `NumberKind` superclass. It could be tested by `isinstance(obj, NumberKind)`.

That test would fail for Add, Mul, etc though because they couldn't inherit `NumberKind` (unless we restrict their args to `NumberKind`).

I guess an abstract superclass might help in some cases with defining or implementing the features expected for an object of a given kind though.

eric-wieser commented:
How do matrices with quaternion elements fit into this design?

oscarbenjamin commented:
> How do matrices with quaternion elements fit into this design?

We'd need to have parametrised kinds like `MatrixKind[NumberKind]`.

JSS95 commented:
Is matrix with non-numeric element used in SymPy?

oscarbenjamin commented:
> Is matrix with non-numeric element used in SymPy?

There is PolyMatrix which has Poly as elements.

Probably in quantum there are matrices of quantum operators etc. Possibly also in vector or somewhere else. Somewhere there are matrices with tuples as elements (although that should really be disallowed).

oscarbenjamin commented:
This has a conflict now

JSS95 commented:
Or maybe we can have `element_kind` property for arrays, `coefficient_kind` and `generator_kind` properties for `Poly`, etc. `Trace(M)` can then refer to `M.element_kind`.

JSS95 commented:
How should we implement parametrised `MatrixKind`? Will it be instance of `MatrixKind` class, or do we need some sort of `ParametrisedKind` class?

JSS95 commented:
Perhaps it must be instance of `MatrixKind` class so that the kinds can be dispatched. I will make a few more commit before merge...

oscarbenjamin commented:
Maybe it isn't necessary to have parametrised kinds and we can just have a way to ask a matrix what the kind of its elements is. A method for doing that would need to be able to recurse through `Integral` etc in the same way as `kind(obj)` though. I'm not sure what's simpler or better.

JSS95 commented:
I would rather choose parametrised kind instead of having another property like 'Integral.element_kind'. Plus, MatrixKind(NumberKind) is good since it resembles the mathematical way of representing a matrix, e.g. 'S.Reals ** (n,m)'.

JSS95 commented:
@oscarbenjamin 
I redesigned the `Kind` class so that it can be parametrised. While doing this, I found out that kind for `Add`, `Mul` and `Pow` needs some more complex dispatching. I have an idea for this, but for now I just removed it to make this PR simple. It will be implemented in succeeding PR.

JSS95 commented:
OK, this is ready to be merged now.

oscarbenjamin commented:
Can there just be singletons for e.g. `NumberKind` and `UndefinedKind`?

There needs to be some specific documentation for `kind`. At the moment it is just in the gotchas section. Each kind should have a clear explanation of what things are and are not included with examples of different sympy classes and expressions.

JSS95 commented:
Oh, okay.

JSS95 commented:
I enhanced the docstrings for kinds

oscarbenjamin commented:
We need a lot more documentation than this. This is a new concept and is being introduced to unmuddy the waters of what sympy objects actually represent but that won't work if it isn't well documented.

I think there should be a specific page in the rst docs that explains the idea of `kind` and also lists the example kinds.

Each kind should have a definition rather than just examples. For example `NumberKind` can be defined as the union of:
```
complex numbers (including real, integer, rational etc)
oo
-oo
zoo
a*oo + b*oo + ... (where a and b are complex numbers)
nan
```
This should also clarify that `Symbol('x')` is `NumberKind` but that `Symbol('x', commutative=False)` is not because `NumberKind` is always commutative. We should give other examples of what are not number kind that might confuse users like quaternions (some might expect these to count as numbers). We need to contrast `NumberKind` with the sympy class `Number` and also with `Expr` so it is clear what the differences are. This should also reference what `is_Number` and `is_number` mean because those are each something different.

This is an area where there is wide potential for confusion. Part of the benefit of introducing kinds as that we can then attach documented definitions. This is especially useful for `NumberKind` which doesn't have any definition elsewhere and is the most common kind. We need the documentation for the kinds themselves to be something that we can reference from other places in the docs so that a function can say that its argument should be NumberKind with a clickable link that leads to a discussion of kinds and a definition of NumberKind.

We need to have examples of what valid operations are for objects of different kinds and so on. It should be worth pointing to the core assumptions as well because the predicates defined there are only really valid for `NumberKind`.

JSS95 commented:
In which docs section do you think the `kind.rst` file should be located? Module reference, or `SymPy Special Topics`?

Upabjojr commented:
> Current design seems very complicated and unintuitive. I think the class design should be like:
> 
> 1. We should have classes for explicit-valued and implicit-valued n-dimensional arrays, all subclassing `NDimArray`. For example, something like `ArraySymbol` should exist.
> 2. Explicit-valued matrix should subclass explicit-valued n-dimensional array. Similar for implicit-valued matrix.
> 3. `MatrixElement` should subclass `ArrayElement`, and perhaps this should subclass `Indexed`.

Please do not change the class structure. The class structure may be complicated and unintuitive, but that's OK: it's internal to SymPy and SymPy users should not deal with it.

Point 2 is very dangerous to apply... the modules are different and you would risk breaking a lot of code. The way operator overloadings and methods work are different. Same reasoning for point 3.

oscarbenjamin commented:
I think both. We can have a page in special topics that introduces the idea of kinds and discusses them at length. Then we can have an autodoc section in modules/core.rst that lists the contents of the module itself and the docstrings (this should link to the more extended discussion in special topics). The docstrings for the kinds themselves should give concise definitions that are clear and complete along with a few examples and counter examples. It should be made clear in each docstring though that there is more information elsewhere (and the html docs should have clickable links to that information).

oscarbenjamin commented:
I also don't think that MatrixKind and ArrayKind should be mixed. If there is to be a unification of those then when that happens the kinds can be changed but for now the kind system should reflect the existing state of the codebase: Matrix and Array are very different things and are not really interchangeable anywhere.

JSS95 commented:
OK, that's good. BTW I found that module reference page is extremely long, untrimmed and complicated. I think I shall make a short PR to break things down before making a detailed doc for kind here..

Upabjojr commented:
Sorry to jump into this _kind_-related discussion too late.

I had a first impression by looking at the code... doesn't it remind of [generics](https://docs.python.org/3/library/typing.html?highlight=generic#typing.Generic)? I mean... since Python 3.5 (with many fixes in Python 3.6), there has been the `typing` module in the standard Python library.

Generics work similarly to C++ templates. What you are trying to do with _kind_ really sounds like generics to me.

What about redefining `Basic` like this:
```python
from typing import TypeVar, Generic

TKind = TypeVar("TKind", covariant=True)

class Basic(Generic[TKind]):
    ...
```
At this point you can really specify the kind in **valid Python code**:
```python
m = Matrix([[1, 2], [3, 4]])
matrix_integral1 = Integral(m, x)
matrix_integral2 = Integral[MatrixKind](m, x)
```
where `matrix_integral1` and `matrix_integral2` are equivalent.

Sorry to jump in like this, but the `typing` module in Python deserves to be analyzed... some nice features are being added to the typesystem of Python.

oscarbenjamin commented:
I think that generics are still tied to the Python type system so unless I misunderstand it isn't possible for different instances of the same class to correspond to different types in difference contexts.

> At this point you can really specify the kind in **valid Python code**:
> 
> ```python
> m = Matrix([[1, 2], [3, 4]])
> matrix_integral1 = Integral(m, x)
> matrix_integral2 = Integral[MatrixKind](m, x)
> ```

And then if I do:
```
matrix_integral3 = Integral(matrix_integral1, x)
```
then the "type" of `matrix_integral3` is `Integral[Integral[MatrixKind]]`? In this context we want it to have the same kind as `matrix_integral1` because it still mathematically represents a matrix.

The point here is that we want the kind of `matrix_integral1` to be `MatrixKind` or possibly `MatrixKind[NumberKind]`. The idea of the kind system is that the fact that it is an instance of the Python class `Integral` is not really relevant if what we want to know is whether it is a matrix or a number so that we can determine if something like `Trace(M)` or `M1 + M2` actually makes sense:
```python
In [1]: M = MatrixSymbol('M', 2, 2)

In [2]: Integral(M, x) + M
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
...
~/current/sympy/sympy/sympy/matrices/expressions/matadd.py in validate(*args)
     91 def validate(*args):
     92     if not all(arg.is_Matrix for arg in args):
---> 93         raise TypeError("Mix of Matrix and Scalar symbols")
     94 
     95     A = args[0]

TypeError: Mix of Matrix and Scalar symbols

In [4]: Trace(Integral(M, x))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-8a9205ab72e5> in <module>
----> 1 Trace(Integral(M, x))

~/current/sympy/sympy/sympy/matrices/expressions/trace.py in __new__(cls, mat)
     31 
     32         if not mat.is_Matrix:
---> 33             raise TypeError("input to Trace, %s, is not a matrix" % str(mat))
     34 
     35         if not mat.is_square:

TypeError: input to Trace, Integral(M, x), is not a matrix
```

eric-wieser commented:
It can't do anything at runtime, but it could be used for mypy typing with a signature like:
```
T = TypeVar("T")

class Integral(Expr[T]):
    def __init__(expr: Expr[T], var: Expr[T]) -> Integral[T]: ...
```
You'd need to tack on some extra tricks to use this at runtime, but it would be cool if we could use the same syntax for both.

Upabjojr commented:
> ```
> matrix_integral3 = Integral(matrix_integral1, x)
> ```
> 
> 
> then the "type" of `matrix_integral3` is `Integral[Integral[MatrixKind]]`? In this context we want it to have the same kind as `matrix_integral1` because it still mathematically represents a matrix.

Generics don't have to follow the expression tree structure. You can define them as you want.

Think about this (imaginary) C++ implementation of `Integral`:

```c++

template <typename Kind>
class Integral {
public:
    Integral(Kind expr, Symbol x) {
        // ...
    }

    Integral(Integral<Kind> nested, Symbol x) {
        // ...
    }

    ...
}
```

Calling the constructor `matrix_integral3 = Integral(matrix_integral1, x)` would resolve to the `typename Kind` of `matrix_integral1`... returning an instance of `Integral<Kind>` identical to `matrix_integral1`.

oscarbenjamin commented:
How would you use generics to determine or specify the type of a Mul like this:
```
x, y = symbols('x, y')
M = MatrixSymb('M', 2, 2)
m = x * y * M
```

Upabjojr commented:
> How would you use generics to determine or specify the type of a Mul like this:
> 
> ```
> x, y = symbols('x, y')
> M = MatrixSymb('M', 2, 2)
> m = x * y * M
> ```

You mean by abolishing `MatMul` and replacing it with something like `Mul[MatrixKind]`?

eric-wieser commented:
You don't even need to abolish `MatMul`, you just make it subclass `Mul[MatrixKind]` instead of the `Mul` it subclasses today (or perhaps, `Mul[MatrixExpr[Number]]` if we want to do away with kinds entirely)

oscarbenjamin commented:
> You mean by abolishing `MatMul` and replacing it with something like `Mul[MatrixKind]`?

My point is how do I have a variadic type like `Mul`, `Add` etc and allow different args to have different generic types?

In the case of `Add` it's not unreasonable to require all args to have the same "type" but for `Mul` I need to allow a mixture.

Does the type `T` in any concrete instantiation have to correspond to an actual type that you can check with `isinstance` or can it just be an arbitrary identifier?

eric-wieser commented:
A quick demo of what `typing.Generic` can already do:
```python
In [24]: from typing import Generic, TypeVar

In [25]: T = TypeVar('T')

In [26]: class Foo(Generic[T]):
    ...:     pass
    ...:

In [28]: class NumberKind: pass

In [29]: FN = Foo[NumberKind]

In [31]: FN.__args__
Out[31]: (__main__.NumberKind,)

In [32]: fn = FN()

In [34]: type(fn)
Out[34]: __main__.Foo

In [36]: fn.__orig_class__ is Foo[NumberKind]
Out[36]: True
```

Upabjojr commented:
> My point is how do I have a variadic type like `Mul`, `Add` etc and allow different args to have different generic types?
> 
> In the case of `Add` it's not unreasonable to require all args to have the same "type" but for `Mul` I need to allow a mixture.

Well, for what I was thinking, `Mul[MatrixKind]` does not mean that the arguments of `Mul` are matrices. It simply means that `Mul[MatrixKind]` behaves as a matrix. Its args can still be of type `Tuple[Expr, ...]`.

If you want to keep the generic parameter to correspond to the types of `.args`, then maybe you can use `Mul[Union[Expr, MatrixExpr]]`.

Upabjojr commented:
Otherwise you can use two generic parameters: one for the kind and one for the type of the `args`: `Mul[MatrixKind, Expr]`.

oscarbenjamin commented:
I've just been playing with it but how do you infer the kind at runtime:
```python
from typing import TypeVar, Generic, get_args
  
T = TypeVar('T')

class MatrixKind:
    pass

class NumberKind:
    pass

class Basic(Generic[T]):
    def __init__(self, *args):
        self.args = args
    def __repr__(self):
        args_str = ', '.join(map(str, self.args))
        return '%s(%s)' % (type(self).__name__, args_str)

class Expr(Basic[T]):
    pass

class Matrix(Expr[MatrixKind]):
    pass

class Integral(Expr[T]):
    def __init__(self, integrand: Expr[T], sym: Expr[NumberKind]):
        super().__init__(integrand, sym)

class Add(Expr[T]):
    def __init__(self, *args: Expr[T]):
        super().__init__(args)

def kind(obj):
    (k,) = get_args(type(obj))
    return k

M = Matrix('M', 2, 2)
N = Add(M, M)

print(kind(M)) # doesn't work
```

eric-wieser commented:
@oscarbenjamin, my code sample above shows how to do that

oscarbenjamin commented:
It doesn't seem to work with a subclass though (in Python 3.8):
```python
In [7]: Expr[MatrixKind]().__orig_class__
Out[7]: __main__.Expr[__main__.MatrixKind]

In [8]: class Matrix(Expr[MatrixKind]): pass

In [9]: Matrix().__orig_class__
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-21b2c4f3664e> in <module>
----> 1 Matrix().__orig_class__

AttributeError: 'Matrix' object has no attribute '__orig_class__'
```

sylee957 commented:
I had some alternative idea like if we had redesigned the `MatMul` completely to always take matrix arguments, and mix with `MatScalarMul` in place, it could better be compatible with typing modules or multipledispatch which doesn't know complicated really types like symmetric and heterogeneous n-ary typings.

eric-wieser commented:
> It doesn't seem to work with a subclass though (in Python 3.8):
> 
> ```python
> In [7]: Expr[MatrixKind]().__orig_class__
> Out[7]: __main__.Expr[__main__.MatrixKind]
> 
> In [8]: class Matrix(Expr[MatrixKind]): pass
> 
> In [9]: Matrix().__orig_class__
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> <ipython-input-9-21b2c4f3664e> in <module>
> ----> 1 Matrix().__orig_class__
> 
> AttributeError: 'Matrix' object has no attribute '__orig_class__'
> ```

Use `__orig_bases__` I think?

oscarbenjamin commented:
> Use `__orig_bases__` I think?

Okay, that works for M:
```python
def kind(obj):
    (b,) = obj.__orig_bases__
    (k,) = get_args(b)
    return k

M = Matrix('M', 2, 2)
N = Add(M, M)

print(kind(M))  # MatrixKind
print(kind(N))  # ~T
```
I didn't want the `kind` of `N` to be `~T` though. The typechecker understands I think that it is `Add[MatrixKind]` but at runtime the type is just `Add`. It would be possible at runtime to check the kinds of the args to `Add` and then generate an instance like `Add[kind(args)](*args)`. That kind of thing would get quite fiddly if we needed it in every `__new__` though.

I think it would be possible to add type hints using generic types to indicate the kinds of the objects. In combination with the Kind system here that would make type hints more meaningful in sympy rather than just the generic `obj: Expr` that we would otherwise have to hint most parameters and returns with.

We would still need the runtime Kind inference implemented in this PR because it isn't possible to access the generic types at runtime even if they are theoretically resolvable by a type checker.

The implication of doing this would be that each `Kind` object would need to be a class rather than an instance. Also in keeping with typing syntax it should probably be square brackets for parametrisation like `MatrixKind[NumberKind]`.

eric-wieser commented:
Yeah, you need a combination of both approaches (class and bases) if you want it to work for both generic and unparameterized types.

oscarbenjamin commented:
> The implication of doing this would be that each `Kind` object would need to be a class rather than an instance. Also in keeping with typing syntax it should probably be square brackets for parametrisation like `MatrixKind[NumberKind]`.

The other implication would be that every `Kind` should have a clear Python class that all its instances belong to e.g. `Expr[NumberKind]` so that how to type-hint any given kind is unambiguous. That would mean that e.g. we shouldn't have any non-Expr objects of kind `NumberKind`.

By the way for anyone interested in using type-hints in sympy I should say that the main obstacle is sympification. Any function that accepts e.g. `Expr` but actually uses `sympify` will in reality accept a huge range of different types: int, float, Rational, complex, mpz, Decimal, str, and so on. I tried to type-hint sympify but actually the main version of the function recursively deconstructs lists:
```
In [90]: sympify([[1, 2], 3])
Out[90]: [[1, 2], 3]
```
There's no way to express that at all with current typing primitives: `Union[T, List[Union[T, List[Union[T, ...`

JSS95 commented:
@oscarbenjamin 

I enhanced docstrings some more.

> We need to have examples of what valid operations are for objects of different kinds and so on. 

Rather than enumerating every available operation to every kind, it will be better to specify the kinds of parameter in the docstring of operator functions and classes.

JSS95 commented:
Plus, I think information about kind itself can be entirely included in docstrings (and thus, in `modules` docs).

The best way to write another dedicated docs for `kind` is to make guide document for (A) Defining new operator class which takes pre-defined kind of objects as arguments (B) Defining completely new objects with new kind. This have to be done in another PR as well, since we don't have any working example for kind-based operator yet.

Upabjojr commented:
> We would still need the runtime Kind inference implemented in this PR because it isn't possible to access the generic types at runtime even if they are theoretically resolvable by a type checker.

Maybe better in a future PR. This PR is already quite long.

Upabjojr commented:
> The implication of doing this would be that each `Kind` object would need to be a class rather than an instance. Also in keeping with typing syntax it should probably be square brackets for parametrisation like `MatrixKind[NumberKind]`.

Not necessarily. Generics can also be objects, the syntax `MatrixKind["hello world!"]` is still valid. Well, of course it depends on how you are going to design the class structure. Usually types are used in generics because you want type-inference to work.

asmeurer commented on doc/src/gotchas.rst:
```diff
@@ -758,6 +758,104 @@ that you want, and they will all be evaluated according to the function.
 
+
+Classification of the objects
+=============================
```
This should go in its own documentation page. It isn't really a gotcha. Furthermore, I think the kind system won't be relevant to most end-users of SymPy.

JSS95 replied:
Where should the document page go? Perhaps in `special topics`?

JSS95 replied:
I moved it to `special topics` because there is nowhere else to put it.

asmeurer commented:
What is the purpose of kind being both a function and an attribute? The only benefit I can see from reading the code is that it `kind()` the function can work on non-SymPy types. But that doesn't seem like a very strong benefit. Everything inside of an expression tree will be Basic, and functions should sympify their arguments anyway.

asmeurer commented:
(Just now catching up on the discussions here). I don't have much experience with typing or languages that have them, so I don't have any thoughts on generics. 

> Currently all Poly have NumberKind as coefficients and generators but in principle it is possible to have e.g. polynomials with matrix coefficients or with matrix generators.

In the case of Poly, we really do want the coefficients to be defined precisely mathematically (perhaps even more so than they currently are). The coefficients of a polynomial should come from a ring, and you should know the properties of that ring (its identity, zero, if it is a field, etc.). I suppose if you had a subset of matrices that defined a commutative ring you could build a coefficient doman out of them (currently Poly doesn't support noncommutative expressions). But this would need to be defined as a ring object that gave all the necessary mathematical properties. 

It's not clear to me how rigorous we want to be with Kind. We apparently don't want Kinds to be sets, since those can be built up symbolically and we want things to be explicit. Do we want them to be rigorous algebraic structures? This is somewhat problematic too, since something like Expr is only vaguely a Ring, since it includes things like infinities and noncommutatives in the same bag of possible objects. Formalizing this mathematically is likely possible, but not simple, and it's not clear what the benefits would be.

oscarbenjamin commented:
> Formalizing this mathematically is likely possible, but not simple, and it's not clear what the benefits would be.

I don't think that the properties of kinds need to be formalised *entirely* but I do think that we should give precise definitions. For example `NumberKind` is a strange sort of union of the extended reals and the extended complex numbers as well as things like `1 + I*oo` and `nan`. There isn't really a comprehensive *mathematical* concept that corresponds to but it is what much of sympy currently works with. It can be defined precisely in the sense of what it includes and what it does not. It is also clearly distinct from a set, boolean, matrix etc. We need explicit definitions so we can clearly exclude the more ambiguous boundary cases like quaternions where someone might think that it counts as a "number" but including it would break a lot of implicit assumptions about the properties of a number.

The emphasis should be on how objects of a given kind can be used in sympy rather than what makes a completely defined mathematical system. Ideally we should aim for those two things to coincide where possible but we have a codebase that works in a certain way and the kind system should reflect that. The kind system gives us a language for describing what *already* happens in a way that is on the one hand documented and on the other computable.

JSS95 commented:
> What is the purpose of kind being both a function and an attribute? The only benefit I can see from reading the code is that it `kind()` the function can work on non-SymPy types. But that doesn't seem like a very strong benefit. Everything inside of an expression tree will be Basic, and functions should sympify their arguments anyway.

I won't mind if we keep only the attribute. @oscarbenjamin , do you agree?

oscarbenjamin commented:
We can remove the kind function. I don't feel that strongly about it but in general I prefer functions to methods and attributes and Basic definitely has way too many methods and attributes. This one is actually worth having though unlike a lot of the others.

oscarbenjamin commented on doc/src/special_topics/classification.rst:
```diff
@@ -0,0 +1,100 @@
+    >>> kind(Integral(A, x))
+    MatrixKind(NumberKind)
+
+This result shows that `Integral(1, x)` is number, and `Integral(A, x)` is array (since matrix is 2-dimensional array) with number element.
```
I guess this should say matrix rather than array.

JSS95 replied:
Oops

oscarbenjamin commented:
We do need more docs than this although potentially that doesn't have to be part of this PR. The explanation of kind in general is very short and there is not much in the way of definitions or examples for each of the particular kinds.

Otherwise I think that this looks good. Does anyone else want to take a look?

Upabjojr commented:
Are there plans to add a dispatcher of methods over the kinds?

For example, we currently call `m.shape` to get the shape of a matrix, if `m` is a matrix. At this point it would be nice to support `Integral(m, x).shape` as well (i.e. `Integral` is of `MatrixKind` if its argument is a matrix).

The question now is:
1. how do we assure that `Integral(m, x).shape` returns the shape of `m`?
2. how do we make sure that calling `Integral(x, x).shape` raises an exception if `x` is neither a matrix nor an array?

JSS95 commented:
@Upabjojr 
We will need singledispatch `shape()` function, which will be implemented in succeeding PR.

moorepants commented on doc/src/special_topics/classification.rst:
```diff
@@ -0,0 +1,100 @@
+numbers, you should consider filtering the arguments with `NumberKind` so that the user
+does not naively pass unsupported objects such as `Integral(A, x)`.
+
+For the performance, set theory is not implemented in kind sysetm. For example,
```
```suggestion
For the performance, set theory is not implemented in kind system. For example,
```

moorepants commented on sympy/core/kind.py:
```diff
@@ -0,0 +1,175 @@
+
+Therefore, we define the "kind" of the object and let the expression
+infer the kind of itself from its arguments. Function and class can
+filer the arguments by their kind, and behave differently according to
```
```suggestion
filter the arguments by their kind, and behave differently according to
```

JSS95 commented:
@oscarbenjamin I believe that this PR can be merged now. Shall I squash the commits?

Upabjojr commented:
> We will need singledispatch `shape()` function, which will be implemented in succeeding PR.

What's the plan to dispatch functions taking more elements having different kind?

For example, `Derivative(m1, m2)` is supposed to be a 4-dim array if both `m1` and `m2` are of `MatrixKind` but it can be interpreted as `MatrixKind` of both `m1` and `m2` are of shape `(k, 1)` (by squashing trivial dimension you get back to a 2-dim array).

oscarbenjamin commented:
> For example, `Derivative(m1, m2)` is supposed to be a 4-dim array if both `m1` and `m2` are of `MatrixKind` but it can be interpreted as `MatrixKind` of both `m1` and `m2` are of shape `(k, 1)` (by squashing trivial dimension you get back to a 2-dim array).

I think that mixing array and matrix like this is a mistake. There should be an explicit function for squashing trivial dimensions to get a matrix from an array.

In any case it would be possible in some way to compute the shape. It's just potentially messy if matrix and array are mixed like this.

Upabjojr commented:
Well, you can derive a matrix by a matrix. The resulting object has 4-dimensions, so it's not a matrix. Some support for handling interactions between `ArrayKind` and `MatrixKind` will be needed. I agree that matrices and arrays should be kept separate.

JSS95 committed:
Introduce kind classification

core/kind module is introduced.
UndefinedKind, NumberKind, BooleanKind, ArrayKind, and MatrixKind
are introduced.
Add, Symbol, Integral, Tr are equipped with kind.
Doc about classifying SymPy objects is introduced.

JSS95 commented:
Squashed the commits.

JSS95 commented:
@oscarbenjamin No suggestion has been for a week (except typo correction), and I don't expect there will be any more. Basic ideas are implemented and documented. We can gradually enhance the docs in succeeding PRs as we make use of kind in other classes. I think this can be merged now and we can start working on applications such as `shape()` function or removing the postprocessors. Any more review, please?

Upabjojr commented:
@JSS95 are you going to implement single dispatching by adding static methods to the `Kind` classes? I see the `.kind` property returns a type, not an object, so is there a way to make use of `functools.singledispatch` by chance?

JSS95 commented:
@Upabjojr `.kind` does not return type. It returns object. I think you are misleaded by `NumberKind` that numeric object's `.kind` returns. In fact, it is a singleton instance of `NumberKind` class. The variable name is immediately redefined in `kind.py` after the class definition.

JSS95 commented:
Also, `functools.singledispatch` will be used for `shape()` function like this: 

```python
@singledispatch
def shape(expr):
    if hasattr(expr, 'shape'):
        # or maybe we need type or kind checking
        return expr.shape
    else:
        raise TypeError

@shape.register(Integral)
def _(expr):
    return expr.function.shape

@shape.register(Derivative)
def _(expr):
    # Infer the shape from arguments
```

Upabjojr commented:
I was trying to see how to implement the single dispatcher, for example:

```python
In [1]: from sympy import *

In [2]: from sympy.matrices import MatrixKind

In [3]: from functools import singledispatch

In [4]: @singledispatch
   ...: def _find_shape(k, x):
   ...:     return ()
   ...:

In [6]: @_find_shape.register
   ...: def _(k: MatrixKind, x):
   ...:     return x.shape
   ...:

In [7]: def shape(x):
   ...:     return _find_shape(x.kind, x)
   ...:

In [8]: shape(eye(3))
Out[8]: (3, 3)

In [9]: x = symbols("x")

In [10]: shape(x)
Out[10]: ()
```

Or do you plan to add a method in `Kind` and `MatrixKind`, e.g.:
```python
class Kind( ... ):
    def _find_shape(self, x):
        return ()


class MatrixKind( ... ):
    ...
    def _find_shape(self, x):
        return x.shape
```
so now the function `shape( )` would be defined as
```python
def shape(x):
    return x.kind._find_shape(x)
```
In both ways you have to pass the element twice.

Do you have a better plan by chance?

Upabjojr commented:
> Also, `functools.singledispatch` will be used for `shape()` function like this:
> 
> ```python
> @singledispatch
> def shape(expr):
>     if hasattr(expr, 'shape'):
>         # or maybe we need type or kind checking
>         return expr.shape
>     else:
>         raise TypeError
> 
> @shape.register(Integral)
> def _(expr):
>     return expr.function.shape
> 
> @shape.register(Derivative)
> def _(expr):
>     # Infer the shape from arguments
> ```

In this example I don't see the need of introducing `.kind`. This could already be implemented without the need of introducing kind. I thought `.kind` would define the dispatcher over the _kind_, not the _class_. It seems like the kind information is not used in your example.

JSS95 commented:
Sorry! My mistake on previous example. We *must* do kind checking in `shape()`.

```
@singledispatch
def shape(expr):
    if isinstance(expr.kind, (MatrixType, ArrayType)):
        return expr.shape
    else:
        raise KindError
```
something like this.

Upabjojr commented:
> Sorry! My mistake on previous example. We _must_ do kind checking in `shape()`.
> 
> ```
> @singledispatch
> def shape(expr):
>     if isinstance(expr.kind, (MatrixType, ArrayType)):
>         return expr.shape
>     else:
>         raise KindError
> ```
> 
> something like this.

I'm still not convinced by this example. `isinstance(expr.kind, (MatrixType, ArrayType))` requires importing from both matrix and array modules in the same file. `@singledispatch` was introduced to avoid importing too many modules in the same file. Why not dispatch over the _kind_ instead?

JSS95 commented:
The primary purpose for kind is to filter the object by its mathatical representation like this.

@oscarbenjamin will have more firm idea on this matter, but in my plan dispatching by *kind* is not used for `shape()`. Instead, it will be used for n-ary operations

JSS95 commented:
If we can have better design by dispatching over *kind*, then it's great!

Upabjojr commented:
> @oscarbenjamin will have more firm idea on this matter, but in my plan dispatching by _kind_ is not used for `shape()`.

I think `shape( )` requires double-dispatching. Dispatch on the class first (for example `shape(Integral( ... )) ==> shape(integral.function)`, then once you've reached the root expression, extract the shape from its _kind_.

If it can help, we have the module `sympy.multipledispatch`.

JSS95 commented:
I'll try that way as well when I start working on `shape()` after this PR gets merged.

Upabjojr commented:
I'm OK with merging this PR. What about the others?

asmeurer commented:
Should we note in the documentation that `kind` is still experimental or in progress?

oscarbenjamin commented:
> Should we note in the documentation that `kind` is still experimental or in progress?

We should definitely make it clear in both release note and docs that it is still experimental.

oscarbenjamin commented:
For implementing something like shape we would need to have dispatching based on regular Python classes because the point is that multiple classes have the same kind. It would be an error to call `shape` for something that is not of kind MatrixKind so dispatching on the kind is useless.

It could be like:
```python
def shape(obj):
    if obj.kind != MatrixKind:
        raise KindError('shape is only defined for matrices')
    return _shape(obj)

@singledispatch
def _shape(obj):
    raise NotImplementedError

@_shape.register(Matrix)
@_shape.register(MatrixExpr)
def _(obj):
    return obj.shape

@_shape.register(Integral)
def _(obj):
    return _shape(obj.args[0])
```

JSS95 committed:
Add experimental note

JSS95 commented:
Made a note that this approach is experimental.

oscarbenjamin commented:
It looks like this is well reviewed and there seems to be agreement about the general approach so I'll merge this.

Actually the tests aren't passed yet but I'll activate the "auto-merge" feature. It will look like I merged it when the tests are passed but it actually just happens automatically.

oscarbenjamin commented:
Thanks @JSS95 for persevering with this. This is good work!

I'm sorry it's slow but all the issues you seem to want to work on go right to the heart of sympy. That's good because those are the most important issues but it also means that it can take longer to get agreement and that there's a higher chance any individual PR might reach a dead end. From my perspective it takes longer to review a PR such like this - the diff here is small which helps a lot but it's still a design change that needs careful consideration. It's definitely worth the time to review a PR like this but I might sometimes hold off for a bit to make sure I have time to think it through properly.

JSS95 commented:
@oscarbenjamin Thanks for merging this, and I'm sorry for being too jumpy. I'll keep it in my mind.

Upabjojr commented:
> For implementing something like shape we would need to have dispatching based on regular Python classes because the point is that multiple classes have the same kind. 

OTOH _MatrixKind_ could be transformed in order to contain the shape information. It would end up being a different design though.

> It would be an error to call `shape` for something that is not of kind MatrixKind so dispatching on the kind is useless.

Well, there's also `ArrayKind`.
